design-documents/distributed-deployment.md

```diff
@@ -0,0 +1,217 @@
+    
+    There are can be a case when extension is not necessarily depends on UI module but needs to be installed on particular instance or instances (for instance, extension that sends an email when the order is placed), in this place it will depend on base module (module that contains business logic) and will be installed on all instances.
+    
+    Pros
+    * Simplifies developer workflow (see below)
+    
+    Cons
+    * Will require us to publish more projects
+    * Potentially some modules that don't depend on UI and depend on base modules may be installed on all instances (need to investigate)
+    * May not work for distributed deployment of services (or one to many relationship, when module need to be installed on multiple instances), see diagram below
+    * How to manage obsolete dependencies (you remove module on once instance, need to remove relevant parts on another)?
+    
+    ![Module dependencies on services](distributed-deployment/module-dependencies-on-services.png)
+
+2. Distribute different projects and introduce new type of configuration for declaring dependencies of packages on instances/services
+
+    Modules would define on which instances they need to be installed and tool would modify composer.json files of these instances to add dependencies to require section.
+    
+    CatalogExtensionAdminUi/composer.json
+    ```
+    {
+        "name": "vendor/catalog-extension-admin-ui",
+        ...
+        "instances": [
+            "admin-ui"
+        ]
+    }
+    ```
+    
+    As when we install extension we use metapackage name (metapackage includes all parts of the extension: AdminUi, Ui, Cron, WebApi) and metapackage is not being installed or saved anywhere, we need to add this information to the composer.json on the instance.
+    
+    admin-ui-instance/composer.json
+    ```
+    {
+        "name": "magento/admin-ui-project",
+        "type": "project",
+        "require": {
+            "vendor/catalog-extension-admin-ui": "*"
+        },
+        "installed-packages": [
+            "vendor/catalog-extension"
+        ]
+    }
+    ```
+    
+    This would allow to understand which packages that are installed on the instance are part of extension we want to remove and remove these packages from all instances.
+    
+    Pros
+     * The same mechanism can be used for distributed deployment of services
+    
+    Cons
+    * Will require us to publish more projects
+    * Need to add new type of configuration
+    
+    Examples of using command
+    
+    ```
+    instance-manager install --package="vendor-name/catalog-customization" --instances "./admin-ui ./ui ./cron ./webapi"
+    instance-manager uninstall --package="vendor-name/catalog-customization" --instances "./admin-ui ./ui ./cron ./webapi"
+    instance-manager update --instances "./admin-ui ./ui ./cron ./webapi"
+    ```
+    
+    Some of corner cases tool need to support
```

Comment:
I don't think that we need to have some mechanism to support these corner cases, it can be done by the composer.